observatory-platform/observatory/platform/utils/url_utils.py

```diff
@@ -14,69 +14,53 @@
-from urllib.parse import ParseResult, urljoin, urlparse
+import logging
 
 import requests
-import tldextract
 import xmltodict
 from importlib_metadata import metadata
 from requests.adapters import HTTPAdapter
 from urllib3.util.retry import Retry
-
-
-def get_url_domain_suffix(url: str) -> str:
-    """Extract a URL composed of the domain name and the suffix of the URL. For example, library.curtin.edu would
-    become curtin.edu
-
-    :param url: a URL.
-    :return: the domain + . + suffix of the URL.
-    """
-
-    result = tldextract.extract(url)
-    return f"{result.domain}.{result.suffix}"
-
-
-def unique_id(string: str) -> str:
-    """Generate a unique identifier from a string.
-
-    :param string: a string.
-    :return: the unique identifier.
-    """
-    return hashlib.md5(string.encode("utf-8")).hexdigest()
-
-
-def is_url_absolute(url: Union[str, ParseResult]) -> bool:
-    """Return whether a URL is absolute or relative.
-
-    :param url: the URL to test.
-    :return: whether the URL is absolute or relative.
-    """
-
-    if not isinstance(url, ParseResult):
-        url = urlparse(url)
-
-    return bool(url.netloc)
-
-
-def strip_query_params(url: str) -> str:
-    """Remove the query parameters from an absolute URL.
-
-    :param url: a URL.
-    :return: the URL with the query parameters removed.
+from tenacity import Retrying, stop_after_attempt, wait_fixed
+from tenacity.wait import wait_base
+
+
+def retry_get_url(
+    url: str, headers: dict = {}, num_retries=3, wait: wait_base = wait_fixed(10), squelch_url: bool = False
+):
+    """Attempts an HTTP GET request inside a retry loop.
+    The request will be retried when the return code is not 200
+
+    :param url: The URL to hit for the GET method
+    :param headers: Any headers to send with the request
+    :param num_retries: The number of times to retry the request before reraising the error
+    :param wait: The Tenacity wait function. If unsupplied, will default to a fixed 10 seconds
+    :param squelch_url: Whether to hide the URL from logs
+    :raises ConnectionError: Raised when the number of retries are exhausted
     """
-    assert is_url_absolute(url), f"strip_query_params: url {url} is not absolute"
-    return urljoin(url, urlparse(url).path)
+    log_url = "" if squelch_url else url
+    for attempt in Retrying(stop=stop_after_attempt(num_retries), wait=wait, reraise=True):
+        with attempt:
+            response = requests.get(url, headers=headers)
+            if response.status_code != 200:
+                logging.info(
+                    f"Downloading events from url: {log_url} returned response code {response.status_code} "
+                    f"attempt: {attempt.retry_state.attempt_number} - idle for: {attempt.retry_state.idle_for}"
+                )
+                raise ConnectionError(
+                    f"Got response code {response.status_code}, response: {response.text}, reason: {response.reason}"
+                )
```

Comment:
It would be good to make the logged message and the ConnectionError have the same message, which would involve adding the response.reason to the error message.

Also the logging error should be more generic - it still has text from Crossref Events, e.g. "Error getting url: " rather than "Downloading events from url"

What happens if requests throws an exception? (see first answer)

https://stackoverflow.com/questions/16511337/correct-way-to-try-except-using-python-requests-module

Reply:
Hi Jamie. I've made some changes to the function that I think make it much cleaner.

requests.get() generally doesn't return an error, so I've added raise_for_status() which will raise an error for any 400/500 return code. This will then be caught, relevant information will be extracted and it will then be reraised. This will be caught by tenacity, which would normally not log an error when retrying, but before_sleep_log() will write the error message to the logger. I've also added a timeout parameter, which will immediately raise an error should the request take too long, which is why the error message changes slightly when a timeout occurs.

I also removed the response.text log as it can be quite lengthy.

The resulting retry logs look like this:
```INFO:__main__:Retrying __main__.retry_get_url in 1.0 seconds as it raised RequestException: Error getting url: http://example.com/frgfds | Got response code: 500 | Reason: Internal Server Error | Attempt: 2 | Idle for: 1.0.```

For timeout:
```INFO:__main__:Retrying __main__.retry_get_url in 1.0 seconds as it raised ReadTimeout: Error getting url: http://example.com/frgfds | Attempt: 1 | Idle for: 0.0.```